### DISCUSSION TITLE
HTTP Client not working (java.lang.ClassNotFoundException)

### DISCUSSION THREAD

**Discussion by Nahuel92** [keycloak/keycloak#30949] [Archive]
Hi guys!

I'm trying to use a REST client from a custom Keycloak SPI (I'm using Keycloak 25.0.1) but without any luck so far. I'd like to ask you if you could give me a clue on how to do it the right way.

I have these dependencies on my `pom.xml`:

```xml
<dependency>
      <groupId>io.quarkus.resteasy.reactive</groupId>
      <artifactId>resteasy-reactive-client</artifactId>
      <version>3.12.0</version>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-rest-client-jackson</artifactId>
      <version>3.12.0</version>
      <scope>provided</scope>
    </dependency>
```

And I declared my API as follows:

```java
@Blocking
@Path("/api")
public interface RestAPI {
    @Path("/users/{id}")
    @GET
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    HttpUserEntity findById(@PathParam("id") String id);
}
```

But I have no idea how to inject it or instantiate it. I tried using:

```java
// Both fail with same exception
final var firstApproach = QuarkusRestClientBuilder.newBuilder().build(RestAPI.class);
final var secondApproach = RestClientBuilder.newBuilder().build(RestAPI.class);
```

Either throws a `java.lang.ClassNotFoundException org.eclipse.microprofile.rest.client.RestClientBuilder`.

With the same configuration, I also tried the following:

```java
final var thirdApproach = ClientBuilder.newClient();
```

But it throws `java.lang.ClassNotFoundException: Provider for jakarta.ws.rs.client.ClientBuilder cannot be found`.

Not sure which approach is the one I should take and documentation isn't clear (and it's outdated, at least Quarkus docs) and there isn't any example of how to use an HTTP client to communicate with a microservice.

I appreciate any clue on how to solve this issue, thanks!

**Comment by hantsy** [comment]
1. Keycloak provides its own `keycloak-admin-client`, it is eaiser to use.
2. If you want to use Quarkus rest client reactive, add `io.quarkus:quarkus-rest-client` to replace your first deps, and the reactive only support the declarative client, check [here](https://github.com/hantsy/quarkus-sandbox/blob/master/restclient-reactive/src/main/java/com/example/demo/PostResourceClient.java)
3. If you want to use classic `ClientBuilder.newClient()`, use legacy resteasy client instead. Add the following deps. 
   ```xml
    <dependency>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-resteasy-client</artifactId>
    </dependency>
    <dependency>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-resteasy-client-jackson</artifactId>
    </dependency>
   ```

**Reply by Nahuel92** [reply]
Hi hantsy!

I tried your second approach without any luck (first one is not what I want to do, I don't need to call another Keycloak instance). This is how I configured my dependencies:

### pom.xml

```xml
<dependencies>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-rest-client</artifactId>
      <version>3.12.0</version>
    </dependency>
</dependencies>
```

When trying to get an instance for my declarative client I still get the `java.lang.ClassNotFoundException`:

```java
RestAPI restAPI = QuarkusRestClientBuilder.newBuilder().build(RestAPI.class); // Exception here
HttpUserEntity entity = restAPI.findById("123");
```

### RestAPI

```java
@Blocking
@Path("/api")
public interface RestAPI {
    @Path("/users/{id}")
    @GET
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    HttpUserEntity findById(@PathParam("id") String id);
}
```

Not sure what is wrong here, my pom contains the dependency that is used at runtime.

Your third approach didn't work either, `java.lang.ClassNotFoundException: Provider for jakarta.ws.rs.client.ClientBuilder cannot be found` is thrown.

### pom.xml

```xml
<dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-resteasy-client</artifactId>
      <version>3.8.5</version>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-resteasy-client-jackson</artifactId>
      <version>3.8.5</version>
    </dependency>
```

### Test code

```java
Client client = ClientBuilder.newClient(); // exception here
```

**Reply by hantsy** [reply]
1. Keycloak admin client usage: https://github.com/hantsy/spring-security-oauth2-resourceserver-example/blob/master/src/test/kotlin/com/example/demo/KeycloakIntegrationTest.kt#L82C1-L87C80
2. Check my example project: https://github.com/hantsy/quarkus-sandbox/blob/master/restclient-reactive
3. For the 3rd solution, check my example codes, https://github.com/hantsy/quarkus-sandbox/blob/master/restclient-jaxrs/src/main/java/com/example/demo/PostResourceClient.java

NOTE: The 2 and 3 just demonstrate the usages of Quarkus Rest Client, you have to handle authentication when connecting to Keycloak Admin Rest API.

**Reply by hantsy** [reply]
And I am not sure your project skeleton is generated from https://code.quarkus.io or Quarkus CLI. 

In the above code fragments, you used different version of these Quarkus deps. In a Quarkus project, you have to use the same quarkus version of the deps and Maven/Gradle quarkus plugins.

**Reply by Nahuel92** [reply]
Hi hansty. I don't need to connect to an external Keycloak instance, I just need to call a microservice (not related to Keycloak).

Here's a PoC project I created to show the issue I'm running into:

https://github.com/Nahuel92/http-user-storage-poc

**Comment by rmegyesi** [comment]
Hi Nahuel!

Using the Microprofile Rest Client is not possible in Keycloak SPIs, because the `quarkus-resteasy-client` extensions are simply not included in the Keycloak build ([Keycloak runtime pom.xml](https://github.com/keycloak/keycloak/blob/main/quarkus/runtime/pom.xml)).

Fortunately there are other options:
* You can acquire a [CloseableHttpClient](https://hc.apache.org/httpcomponents-client-4.5.x/current/httpclient/apidocs/org/apache/http/impl/client/CloseableHttpClient.html) instance from a KeycloakSession in your provider code: `keycloakSession.getProvider(HttpClientProvider.class).getHttpClient()`
* There is a [SimpleHttp](https://github.com/keycloak/keycloak/blob/main/server-spi-private/src/main/java/org/keycloak/broker/provider/util/SimpleHttp.java) utility, that is built on top of CloseableHttpClient
* Use the built-in Java 11+ HTTP client. I usually use this one as it has no dependencies and I can generate my clients with [openapi-generator](https://openapi-generator.tech/docs/generators/java/).

Hope this helps!

**Reply by Nahuel92** [reply]
> Hi Nahuel!
> 
> Using the Microprofile Rest Client is not possible in Keycloak SPIs, because the `quarkus-resteasy-client` extensions are simply not included in the Keycloak build ([Keycloak runtime pom.xml](https://github.com/keycloak/keycloak/blob/main/quarkus/runtime/pom.xml)).
> 
> Fortunately there are other options:
> 
> * You can acquire a [CloseableHttpClient](https://hc.apache.org/httpcomponents-client-4.5.x/current/httpclient/apidocs/org/apache/http/impl/client/CloseableHttpClient.html) instance from a KeycloakSession in your provider code: `keycloakSession.getProvider(HttpClientProvider.class).getHttpClient()`
> * There is a [SimpleHttp](https://github.com/keycloak/keycloak/blob/main/server-spi-private/src/main/java/org/keycloak/broker/provider/util/SimpleHttp.java) utility, that is built on top of CloseableHttpClient
> * Use the built-in Java 11+ HTTP client. I usually use this one as it has no dependencies and I can generate my clients with [openapi-generator](https://openapi-generator.tech/docs/generators/java/).
> 
> Hope this helps!

Hi rmegyesi!!

I'll give that a shot. I tried injecting resteasy's JARs to Keyloack under `/opt/keycloak/libs/deployment`, but I got more errors that I wasn't able to fix.

**Comment by Nahuel92** [comment]
If you are running into the same issue and found this discussion, you can find a working code snippet using `SimpleHttp` here (credits to @ArtyomVancyan):

https://github.com/keycloak/keycloak/discussions/29079#discussioncomment-12040836

That code snippet works really well, and it allows to map the HTTP response to a Java POJO (or record) out of the box.

I'll update my project to include a functional example (i.e., including a dummy HTTP service providing users, using Docker compose, etc.) in the near future.

Having said that, I wish the Keycloak team could enable the official Quarkus HTTP client for devs to use from custom user federation code withouth. It would save us from writing some boilerplate code.

**Reply by jwgmeligmeyling** [reply]
Given the current state of Quarkus integration, the usefulness of the SPI approach is increasingly questionable—especially when implementing something as basic as a REST call requires months of collective effort. It’s surprising how many non-functional solutions are shared as if they’re viable, creating significant confusion. For instance, the Quarkus REST client—reactive or otherwise—consistently fails when used within a Keycloak SPI context, despite community threads and documentation suggesting otherwise. Since the adoption of Quarkus, extending Keycloak through modules has become highly impractical due substantial trial and error. The ecosystem is increasingly constrained by limited, frequently changing, and often poorly documented sets of APIs.

Something definitely needs to be addressed for the SPI to maintain its value.

**Reply by Nahuel92** [reply]
Hi there!

Yes, I never had a chance to update my PoC. I'll do my best to do so these days hopefully. I tested the approach shared above and it works well (and it doesn't require additional setup).

Alternatively, there is a way to make Maven (and Gradle for sure) to copy all the dependencies used by your custom code to the final Jar produced by it.

In case you want to play around with it now, you need to add this in the `build` section of the `pom.xml` file (in case you use Gradle, change it accordingly):

```xml
<build>
<plugin>
	<groupId>org.apache.maven.plugins</groupId>
	<artifactId>maven-shade-plugin</artifactId>
	<version>USE_LATEST_VERSION</version>
	<executions>
		<execution>
			<phase>package</phase>
			<goals>
				<goal>shade</goal>
			</goals>
			<configuration>
				<filters>
					<filter>
						<artifact>*:*</artifact>
						<excludes>
							<exclude>META-INF/*.SF</exclude>
							<exclude>META-INF/*.DSA</exclude>
						</excludes>
					</filter>
				</filters>
			</configuration>
		</execution>
	</executions>
</plugin>
</build>
```

That works well, but the only downside is that the produced uber-Jar will have a larger size. This enables you to use whatever REST client you prefer, not only the ones that are bundled with Quarkus.

Hope this helps. I'll do my best to update my PoC with both approaches, but I can't guarantee when I'll have the time for that.

**Reply by Nahuel92** [reply]
Hi @jwgmeligmeyling

I know it has been a while since you asked for help, but if you are still struggling with this, I just pushed an example that hopefully should unblock you:

https://github.com/Nahuel92/poc-projects/tree/master/keycloak/http-user-storage-poc

It doesn't have a readme file yet, but it's fully Dockerized. As long as you build the jar and then run the `docker-compose.yaml` file, you should be ok.

Good luck!

**Reply by jwgmeligmeyling** [reply]
Hi @Nahuel92 . Thanks for getting back to me and updating your example 😄 We also ended up taking the `SimpleHttp` route. It took a bit of time rewriting our entire integration, but its working again... Nonetheless I am still a bit saddened that none of the standard client libraries seem to work anymore, while documentation suggests otherwise ("_Just_ rewrite to quarkus client")..
